examples/advanced/keycloak-site-authentication/README.md

```diff
@@ -0,0 +1,102 @@
+
+1. Install Python and set up a Virtual Environment,
+```
+python3 -m venv nvflare-env
+source nvflare-env/bin/activate
+```
+2. Install NVFlare
+```
+pip install nvflare
+```
+3. The example is part of the NVFlare source code. The source code can be obtained like this,
+```
+git clone https://github.com/NVIDIA/NVFlare.git
+```
+4. TLS requires domain names. Please add the following line in the `/etc/hosts` file,
+```
+127.0.0.1	server1
+```
+
+### Setup
+
+```
+nvflare poc prepare -i project.yml -c site_a site_b
+```
+
+All the startup kits will be generated in this folder,
+```
+/tmp/nvflare/poc/keycloak-site-authentication/prod_00
+```
+
+### Set up KeyCloak
+
+Download and install the KeyCloak following the instruction guide from https://www.keycloak.org/getting-started/getting-started-zip. Key steps for the installation:
+
+* Start the KeyCLoak by running "bin/kc.sh start-dev"
+* Set up the realm called "myrealm"
+* Setup the user "myuser@example.com"
+* Add the client "myclient"
+* The KeyCLoak will be running at http://localhost:8080/
+
+### Set up Client side authentication requirement
+
+```
+cp -r site/local/* /tmp/nvflare/poc/keycloak-site-authentication/prod_00/site_a/local/* 
+```
+This will set up site_a with additional user security authentication to the KeyCloak.
+
+### Set up Admin user authentication
+
+```
+cp -r admin/local/* /tmp/nvflare/poc/keycloak-site-authentication/prod_00/myuser@example.co/local/* 
+```
+This will set up admin user myuser@example.com to acquire KeyCloak access_token when starting the admin tool.
+
+
+### Starting NVFlare
+
+This script will start up the server and 2 clients,
+```
+nvflare poc start
+```
+
+### Logging with Admin Console
+
+Start the myuser@example.com admin tool and login as the `myuser@example.com` user. Also provide the password to the KeyCloak:
+
+```
+cd /tmp/nvflare/poc/job-level-authorization/prod_00/myuser@example.com
+./startup/fl_admin.sh
+```
+
+At the prompt, enter the user email `myuser@example.com`, and then provide the password to `site_a` KeyCloak.
+
+
+### User authentication when running jobs
+
+With this system set up, the `aite_a` will require additional user authentication when submitting and running a job. `site_b` does not have this additional security requirement. Any admin user can submit and run the job.
+
+Let's choose the `hello-numpy-sag` job from the `hello-world` examples. For demonstrating purpose, let's change the `min_clients` in the job meta.json to 1.  
+
+#### Authenticated admin user
+
+* `myuser@example.com` corrected authenticated to `site_a` KeyCloak system. When he submitting the job, the `hello-numpy-sag` job is successfully running on both `site_a` and `site_b`.
+
+#### Un-authenticated admin user
+
+* If the `myuser@example.com` admin user provides the wrong password, or for some reason KeyCloak system is not available when starting the admin tool, or submitting the job, the `hello-numpy-sag` job won't be able to run the `site_a`. 
```

Comment:
If the user fails the 3rd-party authentication, the user shouldn't be able to log in to the admin console.
If the user skips the 3rd-party authentication (by not installing the AdminAuth), the user submitted jobs won't be run by site-a (we should test this).

Reply:
I am not sure if NVFlare should dictate that if the admin user failed one 3rd-party authentication, then the admin tool won't be able to start or not. This probably be left to the customers to decide. 

Also one admin user could potentially login to a system with site-1, site-2, site-3 three sites. But this admin may only have credentials for site-1, and site-2. And in this case, the admin may only want to submit jobs to site-1 and site-2.

Reply:
Added instruction comment for the customers on how to change to the behavior of stopping the admin tool if admin failed to be authenticated to the 3rd-party system.